File: 07-week-6--foundational-data-structures/05-bonus--build-a-doubly-linked-list/solutions/doubly_linked_list.js
class DoublyLinkedList {
  constructor(head = null) {
    this.head = head;
  }

  iterate(callback) {
    let count = 0;
    let temp = this.head;

    while (temp !== null) {
      const result = callback(temp, count);

      if (result === true) {
        return temp;
      }

      ++count;
      temp = temp.next;
    }

    return this.head;
  }

  // print each node's value on its own line
  // use your iterate method to be DRY! Don't get caught in the code rain, brrr.
  print() {
    this.iterate(node => console.log(node.value));
  }

  // find the node with the target value and return it
  // if not found return null, use your iterate method to be DRY!
  find(target) {
    let result = null;

    this.iterate(node => {
      if (node.value === target) {
        result = node;

        return true;
      }
    });

    return result;
  }

  // add the node to the start of the list, no nodes should be removed
  // MODIFY TO UPDATE PREV
  addFirst(node) {
    node.next = this.head;

    if (this.head !== null) {
      this.head.prev = node;
    }
    
    this.head = node;
  }

  // add node to end of list, no nodes should be removed
  // you may wish to use the iterate method
  // MODIFY TO UPDATE PREV
  addLast(node) {
    if (this.head === null) {
      this.head = node;
      return;
    }

    this.iterate(currNode => {
      if (currNode.next === null) {
        currNode.next = node;
        node.prev = currNode;
        return true;
      }
    });
  }

  // remove the first Node in the list and update head
  // and return the removed node
  // MODIFY TO UPDATE PREV
  removeFirst() {
    const oldHead = this.head;

    if (this.head !== null) {
      this.head = this.head.next;
    }

    if (this.head !== null) {
      this.head.prev = null;
    }

    return oldHead;
  }

  // remove the tail node, iterate may be helpful
  // return the node you just removed
  removeLast() {
    if (this.head === null || this.head.next === null) {
      return this.removeFirst();
    }

    let oldTail = null;

    this.iterate(node => {
      if (node.next.next === null) {
        oldTail = node.next;
        node.next = null;
        return true;
      }
    });

    return oldTail;
  }

  // replace the node at the given index with the given node
  // MODIFY TO UPDATE PREV
  replace(idx, node) {
    if (idx === 0) {
      this.removeFirst();
      this.addFirst(node);
      return node;
    }

    this.iterate((currNode, count) => {
      if (count === idx - 1) {
        node.next = currNode.next.next;
        currNode.next = node;
        node.prev = currNode;

        if (currNode.next.next !== null) {
          currNode.next.next.prev = node;
        } 

        return true;
      }
    });

    return node;
  }

  // insert the node at the given index
  // no existing nodes should be removed or replaced
  // MODIFY TO UPDATE PREV
  insert(idx, node) {
    if (idx === 0) {
      this.addFirst(node);
      return;
    }

    this.iterate((currNode, count) => {
      if (count === idx - 1) {
        const oldNext = currNode.next;
        currNode.next = node;
        node.next = oldNext;
        node.prev = currNode;

        if (oldNext !== null) {
          oldNext.prev = node;
        }

        return true;
      }
    });
  }

  // remove the node at the given index, and return it
  // MODIFY TO UPDATE PREV
  remove(idx) {
    if (idx === 0) {
      return this.removeFirst();
    }

    let oldNode = null;

    this.iterate((node, count) => {
      if (count === idx - 1) {
        oldNode = node.next;
        node.next = node.next.next;

        if (node.next !== null) {
          node.next.prev = node;
        }

        return true;
      }
    }); 

    return oldNode;
  }

  clear() {
    this.head = null;
  }
}

class Node {
  constructor(value = null, next = null, prev = null) {
    this.value = value;
    this.next = next;
    this.prev = prev;
  }
}

if (require.main === module) {
  let node = new Node('one');
  node.prev = new Node('zero');

  console.log("NODE TESTING PREV VALUES");
  console.log(`Node prev value is zero: ${node.prev.value}`);
  console.log(`Node prev.prev is null: ${node.prev.prev}`);
  console.log("---------------------------------------------------------");

  node = new Node('one');
  let list = new DoublyLinkedList(node);
  list.addFirst(new Node('zero'));

  console.log("TEST PREV WHEN CALLING ADD FIRST");
  console.log(`Node prev value is zero: ${node.prev.value}`);
  console.log(`Node prev.next value is one: ${node.prev.next.value}`);
  console.log("---------------------------------------------------------");

  let lastNode = new Node('last');
  list.addLast(lastNode);
  console.log("TEST PREV WHEN CALLING ADD LAST");
  console.log(`Last node prev value is one: ${lastNode.prev.value}`);
  console.log("---------------------------------------------------------");

  console.log("TEST PREV WHEN CALLING REMOVE FIRST");
  list.removeFirst();
  console.log(`Head node's prev should be null: ${list.head.prev}`);
  console.log("---------------------------------------------------------");

  // one -> last
  list.replace(0, new Node('zero'));
  console.log("TEST PREV WHEN REPLACING HEAD");
  console.log(`Head node's prev is null: ${list.head.prev}`);

  // zero -> last
  list.replace(1, new Node('one'));
  console.log("TEST PREV WHEN REPLACING TAIL");
  console.log(`Tail node's prev is zero: ${list.head.next.prev.value}`);

  let head = new Node('zero');
  let one = new Node('one');
  let two = new Node('two');
  head.next = one
  one.next = two
  one.prev = head
  two.prev = one
  list = new DoublyLinkedList(head);

  // zero -> one -> two
  replacing = new Node('replacing');
  list.replace(1, replacing)
  // zero -> replacing -> two
  console.log("TEST PREV WHEN REPLACING MIDDLE NODE");
  console.log(`Middle node prev value is zero: ${replacing.prev.value}`);
  console.log(`Last node prev value is replacing: ${two.prev.value}`);
  console.log("---------------------------------------------------------");

  // zero -> replacing -> two
  node = new Node('insert');
  list.insert(0, node);
  // insert -> zero -> replacing -> two
  console.log("TEST PREV WHEN INSERTING 0TH NODE");
  console.log(`Head node prev is null: ${list.head.prev}`);
  console.log(`Head node next.prev value is insert: ${list.head.next.prev.value}`);

  node = new Node('at tail');
  list.insert(4, node);
  // insert -> zero -> replacing -> two -> at tail
  console.log("TEST PREV WHEN INSERTING TAIL NODE");
  console.log(`Tail node prev is two: ${node.prev.value}`);

  node = new Node('in middle');
  list.insert(2, node);
  // insert -> zero -> in middle -> replacing -> two -> at tail
  console.log("TEST PREV WHEN INSERTING IN MIDDLE");
  console.log(`Middle node prev value is zero: ${node.prev.value}`);
  console.log(`Middle node next.prev is in middle: ${node.next.prev.value}`);
  console.log("---------------------------------------------------------");

  // insert -> zero -> in middle -> replacing -> two -> at tail
  list.remove(0);
  // zero -> in middle -> replacing -> two -> at tail
  console.log("TEST PREV WHEN REMOVING 0TH NODE");
  console.log(`Head node's prev value is null: ${list.head.prev}`);

  list.remove(2);
  // zero -> in middle -> two -> at tail
  console.log("TEST PREV WHEN REMOVING IN MIDDLE");
  console.log(`Node at index 2 prev's value is in middle: ${list.head.next.next.prev.value}`);

  list.remove(3);
  // zero -> in middle -> two 
  console.log("TEST PREV WHEN REMOVING TAIL");
  console.log(`Tail node's prev should still point to node with value in middle: ${list.head.next.next.prev.value}`);
  console.log("---------------------------------------------------------");
}

module.exports = {
  Node, DoublyLinkedList
};

File: 07-week-6--foundational-data-structures/05-bonus--build-a-doubly-linked-list/solutions/doubly_linked_list.rb
class DoublyLinkedList
  attr_accessor :head

  def initialize(head = nil)
    @head = head
  end

  def iterate
    count = 0
    temp = @head

    until temp.nil?
      yield(temp, count)      
      temp = temp.next_node
      count += 1
    end

    @head
  end

  def print
    iterate { |node| puts node.value }
  end

  def find(target)
    iterate do |node|
      return node if node.value == target
    end

    nil
  end

  # MODIFY TO UPDATE PREV_NODE
  def add_first(node)
    node.next_node = @head
    @head.prev_node = node unless @head.nil?
    @head = node
  end

  # MODIFY TO UPDATE PREV_NODE
  def add_last(node)
    if @head.nil?
      @head = node
      return
    end

    iterate do |curr_node|
      if curr_node.next_node.nil?
        curr_node.next_node = node
        node.prev_node = curr_node
        return
      end
    end
  end

  # MODIFY TO UPDATE PREV_NODE
  def remove_first
    old_head = @head
    @head = @head.next_node unless @head.nil?
    @head.prev_node = nil unless @head.nil?
    old_head
  end

  def remove_last
    return remove_first if @head.nil? || @head.next_node.nil?

    iterate do |node|
      if node.next_node.next_node.nil?
        old_tail = node.next_node
        node.next_node = nil
        return old_tail
      end
    end
  end

  # MODIFY TO UPDATE PREV_NODE
  def replace(idx, node)
    if idx.zero?
      remove_first
      add_first(node)
      return node
    end

    iterate do |curr_node, count|
      if count == idx - 1
        node.next_node = curr_node.next_node.next_node
        curr_node.next_node = node
        node.prev_node = curr_node

        unless curr_node.next_node.next_node.nil?
          curr_node.next_node.next_node.prev_node = node
        end

        return node
      end
    end
  end

  # MODIFY TO UPDATE PREV_NODE
  def insert(idx, node)
    if idx.zero?
      add_first(node)
      return
    end

    iterate do |curr_node, count|
      if count == idx - 1
        old_next = curr_node.next_node
        curr_node.next_node = node
        node.next_node = old_next
        old_next.prev_node = node unless old_next.nil?
        node.prev_node = curr_node

        return
      end
    end
  end

  # MODIFY TO UPDATE PREV_NODE
  def remove(idx)
    if idx.zero?
      return remove_first
    end

    iterate do |node, count|
      if count == idx - 1
        old_node = node.next_node
        node.next_node = node.next_node.next_node
        node.next_node.prev_node = node unless node.next_node.nil?

        return old_node
      end
    end
  end

  def clear
    @head = nil
  end
end

class Node
  attr_accessor :value, :next_node, :prev_node

  def initialize(value = nil, next_node = nil, prev_node = nil)
    @value = value
    @next_node = next_node
    @prev_node = nil
  end
end

if __FILE__ == $PROGRAM_NAME
  node = Node.new('one')
  node.prev_node = Node.new('zero')

  puts "NODE TESTING PREV VALUES"
  puts "Node prev value is zero: #{node.prev_node.value}"
  puts "Node prev.prev is null: #{node.prev_node.prev_node}"
  puts "---------------------------------------------------------"

  node = Node.new('one')
  list = DoublyLinkedList.new(node)
  list.add_first(Node.new('zero'))

  puts "TEST PREV WHEN CALLING ADD FIRST"
  puts "Node prev value is zero: #{node.prev_node.value}"
  puts "Node prev.next value is one: #{node.prev_node.next_node.value}"
  puts "---------------------------------------------------------"

  last_node = Node.new('last')
  list.add_last(last_node)
  puts "TEST PREV WHEN CALLING ADD LAST"
  puts "Last node prev value is one: #{last_node.prev_node.value}"
  puts "---------------------------------------------------------"

  puts "TEST PREV WHEN CALLING REMOVE FIRST"
  list.remove_first
  puts "Head node's prev should be nil: #{list.head.prev_node}"
  puts "---------------------------------------------------------"

  # one -> last
  list.replace(0, Node.new('zero'))
  puts "TEST PREV WHEN REPLACING HEAD"
  puts "Head node's prev is null: #{list.head.prev_node}"

  # zero -> last
  list.replace(1, Node.new('one'))
  puts "TEST PREV WHEN REPLACING TAIL"
  puts "Tail node's prev is zero: #{list.head.next_node.prev_node.value}"

  head = Node.new('zero')
  one = Node.new('one')
  two = Node.new('two')
  head.next_node = one
  one.next_node = two
  one.prev_node = head
  two.prev_node = one
  list = DoublyLinkedList.new(head)

  # zero -> one -> two
  replacing = Node.new('replacing')
  list.replace(1, replacing)
  # zero -> replacing -> two
  puts "TEST PREV WHEN REPLACING MIDDLE NODE"
  puts "Middle node prev value is zero: #{replacing.prev_node.value}"
  puts "Last node prev value is replacing: #{two.prev_node.value}"
  puts "---------------------------------------------------------"

  # zero -> replacing -> two
  node = Node.new('insert')
  list.insert(0, node)
  # insert -> zero -> replacing -> two
  puts "TEST PREV WHEN INSERTING 0TH NODE"
  puts "Head node prev is null: #{list.head.prev_node}"
  puts "Head node next.prev value is insert: #{list.head.next_node.prev_node.value}"

  node = Node.new('at tail')
  list.insert(4, node)
  # insert -> zero -> replacing -> two -> at tail
  puts "TEST PREV WHEN INSERTING TAIL NODE"
  puts "Tail node prev is two: #{node.prev_node.value}"

  node = Node.new('in middle')
  list.insert(2, node)
  # insert -> zero -> in middle -> replacing -> two -> at tail
  puts "TEST PREV WHEN INSERTING IN MIDDLE"
  puts "Middle node prev value is zero: #{node.prev_node.value}"
  puts "Middle node next.prev is in middle: #{node.next_node.prev_node.value}"
  puts "---------------------------------------------------------"

  # insert -> zero -> in middle -> replacing -> two -> at tail
  list.remove(0)
  # zero -> in middle -> replacing -> two -> at tail
  puts "TEST PREV WHEN REMOVING 0TH NODE"
  puts "Head node's prev value is nil: #{list.head.prev_node}"

  list.remove(2)
  # zero -> in middle -> two -> at tail
  puts "TEST PREV WHEN REMOVING IN MIDDLE"
  puts "Node at index 2 prev's value is in middle: #{list.head.next_node.next_node.prev_node.value}"

  list.remove(3)
  # zero -> in middle -> two 
  puts "TEST PREV WHEN REMOVING TAIL"
  puts "Tail node's prev should still point to node with value in middle: #{list.head.next_node.next_node.prev_node.value}"
  puts "---------------------------------------------------------"
end

# Please add your pseudocode to this file
# And a written explanation of your solution
